### ISSUE TITLE
Support installing via git https URL from any domain name

### ISSUE THREAD

**Issue by silverwind** [npm/npm#19788]
Currently, `npm` does not support installation of modules from git services hosted on private domain names. That includes both Github for Enterprise on custom domains as well as instances of gitlab, gitea, gogs, bitbucket and many others, basically anything hosted on a custom domain name.

`git` supports cloning from any URL, so I don't see why `npm` should not. This works:
````sh
git clone https://git.mydomain.com/user/repo.git
````
This fails:
````sh
npm install https://git.mydomain.com/user/repo.git
````

The root of the issue is that [hosted-git-info](https://github.com/npm/hosted-git-info) acts as a whitelist of domains and if one tries to install from another domain, npm fails with a nondescript error that can be traced back to that module. I have [this PR](https://github.com/npm/hosted-git-info/pull/30) up which should eventually allow npm to install packages from any domain name.

**Comment by chris-dura** [comment]
@silverwind So, obviously you reference installing via an `http(s)://` URL directly, but just as an fyi, our GitLab Enterprise instance allows us to install using a slightly different format. We have 2FA enabled, so it requires SSH to be used. From the [docs](https://docs.npmjs.com/cli/install#synopsis).
```bash
npm install <git-host>:<git-user>/<repo-name>
npm install <git repo url>
```

We were able to actually install our repos like this:
```bash
npm install git+ssh://git@gitlab.mydomain.com:user/repo.git
```

This did not; however, I imagine it would if we allowed "Basic Auth" on our GitLab Enterprise:
```bash
npm install git+https://git@gitlab.mydomain.com:user/repo.git
```

And, of course, these didn't work, giving "no package.json" errors. The error is a bit misleading, but probably *technically* correct since it's probably not properly resolving the URL.
```bash
npm install https://gitlab.mydomain.com/user/repo
npm install https://gitlab.mydomain.com/user/repo.git
```

**Comment by silverwind** [comment]
Indeed, if I prepend `git+` on the HTTPS URL it works (I run gitea which accepts basic auth). Still, I think npm should support any plain https URLs which is cloneable, as stated in the `npm install <git repo url>` synopsis.

**Comment by silverwind** [comment]
> And, of course, these didn't work, giving "no package.json" errors. The error is a bit misleading, but probably technically correct since it's probably not properly resolving the URL.

I see no correctness in that behaviour. If git can clone it, npm can get the files. As mentioned, the root issue is that npm only allows plain https URLs from Github, Gitlab and Bitbucket, nothing else.

**Comment by chris-dura** [comment]
> I see no correctness in that behaviour.

Sorry, I meant the error message was sort of correct, not the behavior.

**Comment by silverwind** [comment]
Bump! Many people are currently looking for GitHub alternatives now that it has been sold out. Would appreciate if someone from the npm team could have a look at https://github.com/npm/hosted-git-info/pull/30.
